components/raftstore/src/store/fsm/peer.rs

```diff
@@ -688,53 +689,51 @@ where
                 }
             }
         }
-        // Propose batch request which may be still waiting for more raft-command
-        if self.ctx.sync_write_worker.is_some() {
-            self.propose_batch_raft_command(true);
-        } else {
-            self.propose_batch_raft_command(false);
-            self.check_batch_cmd_and_proposed_cb();
-        }
+        self.on_loop_finished();
     }
 
-    fn propose_batch_raft_command(&mut self, force: bool) {
-        if self.fsm.batch_req_builder.request.is_none() {
-            return;
-        }
-        if !force
-            && self.ctx.cfg.cmd_batch_concurrent_ready_max_count != 0
-            && self.fsm.peer.unpersisted_ready_len()
-                >= self.ctx.cfg.cmd_batch_concurrent_ready_max_count
-        {
-            return;
+    #[inline]
+    #[allow(unused_mut, clippy::redundant_closure_call)]
+    fn on_loop_finished(&mut self) {
+        let mut ready_concurrency = self.ctx.cfg.cmd_batch_concurrent_ready_max_count;
+        let mut should_propose = self.ctx.sync_write_worker.is_some()
+            || ready_concurrency == 0
+            || self.fsm.peer.unpersisted_ready_len() < ready_concurrency;
+        (|| {
+            fail_point!("force_delay_propose_batch_raft_command", |_| {
+                should_propose = false
+            })
+        })();
+        // Propose batch request which may be still waiting for more raft-command
+        if should_propose {
+            self.propose_pending_batch_raft_command();
+        } else if let Some(cmd) = self.fsm.batch_req_builder.request.take() {
```

Comment:
This should be executed when it has not been executed and there is proposed callback.

Reply:
I don't know what you mean. But I fixed some inconsistency with old behavior.